llvm/tools/llvm-exegesis/lib/RISCV/RISCVCounters.cpp

```diff
@@ -0,0 +1,73 @@
+//===-- RISCVCounters.cpp ---------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines RISC-V perf counters.
+//
+//===----------------------------------------------------------------------===//
+
+#include "RISCVCounters.h"
+
+namespace llvm {
+namespace exegesis {
+
+// This implementation of RISCV target for Exegesis doesn't use libpfm
+// and provides manual implementation of performance counters.
+
+inline uint64_t getRISCVCpuCyclesCount() {
+#ifdef __riscv
+  uint64_t Counter;
+  asm("csrr %0, cycle" : "=r"(Counter)::"memory");
```

Comment:
It is incorrect on rv32. See also https://github.com/google/benchmark/blob/8e1823d6f59c1d0fdc084fd903c989e6816ea097/src/cycleclock.h#L192-L214.

BTW, as cycle counter is unavailable since Linux 6.6, I think libpfm is still needed.

See also
https://github.com/google/benchmark/commit/99bdb2127d1fa1cff444bbefb814e105c7d20c45

Reply:
Libpfm support looks redundant for this patch. We can do it in a separate patch. 
Added support for rv32.

Reply:
I don't think it is correct to use `mtime`. The measurement result may be affected by dynamic frequency scaling.

Reply:
If `libpfm` works on the platforms that you're interested in, or even some of them, I think we should prefer that. You get all the perf counter stuff essentially for "free" since it's already there.

Adding another counter using inline assembly could be done later in a future PR (unless there is no libpfm/Linux perf support).

Reply:
Now `rdcycle` instructions used, instead of `rdtime`. Current we don`t have libpfm support for RISC-V

Reply:
Addressed